packages/backend/src/core/SequenceProcessor.ts

```diff
@@ -0,0 +1,120 @@
+  startFrom: number
+  batchSize: number
+  getLast: (previousLast: number) => Promise<number>
+  processRange: (
+    // [from, to] <- ranges are inclusive
+    from: number,
+    to: number,
+    trx: Knex.Transaction,
+  ) => Promise<void>
+  scheduleIntervalMs?: number
+}
+
+export const ALL_PROCESSED_EVENT = 'All processed'
+
+const HOUR = 1000 * 60 * 60
+
+export class SequenceProcessor extends EventEmitter {
+  readonly id: string
+  private readonly processQueue: TaskQueue<void>
+  private readonly scheduleInterval: number
+  private readonly logger: Logger
+  private processedAll = false
+  private refreshId: NodeJS.Timer | undefined
+
+  constructor(private readonly opts: SequenceProcessorOpts) {
+    super()
+
+    assert(opts.batchSize > 0)
+
+    this.id = this.opts.id
+    this.logger = this.opts.logger.for(`${SequenceProcessor.name}.${this.id}`)
+    this.processQueue = new TaskQueue<void>(
+      () => this.process(),
+      this.logger.for('updateQueue'),
+      {
+        shouldRetry: Retries.exponentialBackOff(100, {
+          maxDistanceMs: 3_000,
+          maxAttempts: 10,
+        }),
+      },
+    )
+    this.scheduleInterval = opts.scheduleIntervalMs ?? HOUR
+  }
+
+  public start(): void {
+    this.logger.info('Started')
+    this.processQueue.addIfEmpty()
+    this.refreshId = setInterval(
+      () => this.processQueue.addIfEmpty(),
+      this.scheduleInterval,
+    )
+  }
+
+  public async stop(): Promise<void> {
+    this.logger.info('Stopping')
+    clearInterval(this.refreshId)
+    await this.processQueue.waitTilEmpty()
+  }
+
+  hasProcessedAll(): boolean {
+    return this.processedAll
+  }
+
+  private async process(): Promise<void> {
+    this.logger.debug('Processing started')
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    processing: while (true) {
+      const lastProcessed = (await this.opts.repository.getById(this.id))
+        ?.lastProcessed
+      this.logger.debug('Calling getLast')
+      const to = await this.opts.getLast(lastProcessed ?? this.opts.startFrom)
+
+      if ((lastProcessed ?? this.opts.startFrom) === to) {
+        break processing
+      }
+      // to avoid processing lastSynced multiple times we need to increment it by one
+      let from = lastProcessed ? lastProcessed + 1 : this.opts.startFrom
+
+      assert(
+        from <= to,
+        `getLast returned sequence member that was already processed. from=${from}, to=${to}`,
+      )
+
+      for (; from <= to; from += this.opts.batchSize) {
+        await this.processRange(
+          from,
+          Math.min(from + this.opts.batchSize - 1, to),
+        )
+      }
+    }
+
+    this.logger.debug('Processing finished')
+    this.processedAll = true
+    this.emit(ALL_PROCESSED_EVENT)
+  }
+
+  private async processRange(from: number, to: number) {
+    this.logger.debug('Processing range started', { from, to })
+    await this.opts.repository.runInTransaction(async (trx) => {
```

Comment:
Question. Is it ok to have such long running transactions? Is this non-blocking for other things in the database? I have a feeling it's fine, but another feeling that it might cause problems.

Reply:
It is a very good question, I thought about it, but did not want to complicate it until I'm sure this is an issue. I added a task for looking into it - ideally it would not start a transaction until a first call to database is issued, but it might not be the case - will make it work so anyway when looking into it.